### ISSUE TITLE
Safari: Upload test doesn't work on mpot branch

### ISSUE THREAD

**Issue by adolfintel** [adolfintel/speedtest#183]
At the end of November 2018, Apple broke the implementation of CORS headers on Safari. As a result, the upload test doesn't work on the mpot branch.
This affects Safari on macOS and all browsers on iOS, but is limited to the mpot branch.
**Important: this issue does not affect the regular version of the speedtest, only the mpot branch!**

**Comment by karthikkp** [comment]
This happens on the node branch also. May be you can point to the problem in the code and I can raise a PR

**Comment by adolfintel** [comment]
@karthikkp Are you sure? The node branch shouldn't use CORS at all. Do you have the server on a different domain than the UI or something?

@dunklesToast Can you check this? Thanks

**Comment by karthikkp** [comment]
@adolfintel I am sure. I don't have the server on a different domain as well.

**Comment by adolfintel** [comment]
@karthikkp Ok that's weird. Tom says he's working on a rewrite of the node version (which is currently pretty out of date) so you might want to wait for the new version (he says it will be ready in the next 1-2 days), or switch to the regular version.

**Comment by aficustree** [comment]
I'm not sure this has anything to do with CORS. I took the upload code in the mpot branch and sync'd it back up with the master branch and the issue persists. Walking through a debugger, I think it's actually a binding issue. The `totLoaded` variable increments properly but is zero each time it's called through the set interval callback. There is a bind(this) which looks proper so I think it should have the right reference but safari seems to have lost the scope.

**Comment by adolfintel** [comment]
@aficustree That sucks, last time I checked, the master branch worked fine on safari, but the mpot branch spewed CORS errors in the console. Can you check the demo at [speedtest.fdossena.com](http://speedtest.fdossena.com) and see if the master branch is now affected too?

I don't have any apple device to test this on so if you could [contact me on telegram](https://t.me/dosse91) and help me sort this out, it would be great.

**Comment by aficustree** [comment]
@adolfintel , looks like I'm wrong but the issue is quite odd. I've tested a number of scenarios.

1. MPOT+Node+Classic when behind anything that adds a CORS header results in no upload functionality on Safari. The CORS header is added by default in MPOT which is why that's seen as the most commonly impacted version. 
2. When provided a CORS header, the issue arrises. It seems to be here:

- the`xhr[i].upload.onprogress` function is properly triggered and increments the `totLoaded` variable. the totLoaded variable is seen within the ulTest closure
- the setInterval thread is run by 'window'. There's a lot of random hits that Safari has been playing with additional isolation of window when in CORS mode due to concerns around Spectre attacks. It's all a bit hard to follow but I think that in CORS mode the POST method triggers a shadow copy of the closure variables within the onprogress callback and thus window is isolated from seeing the updates.

**Comment by adolfintel** [comment]
@aficustree That's weird but it could be it. No idea why it doesn't happen with the download test though.
Anyway, we could try working around the issue by having a global variable totloaded_safari that's incremented instead. Can you try doing that?

**Comment by aficustree** [comment]
I tried kicking totLoaded up to a global variable and it appears that nothing that's modified while inside a POST event can get a variable back to the window. If you watch the debugger, the callbacks in the POST event actually loses the global vars scope and instead creates its own copy. I think the only way this continues to work in safari is if a timer object is the one that calls a series of post actions and does the measurement and/or more server-side reporting of timing is developed.

**Comment by adolfintel** [comment]
That's terrible, web workers were already limited in how they can communicate with window, but this is completely incorrect and against the spec.
I guess we could, as a last resort, get rid of the worker altogether but having the test on a separate thread is a better solution.

Just as a last attempt, what happens if we try to report the status from inside the timer in ulTest? In theory, totLoaded should be visible to it, and once postMessage is called, the string shouldn't be touched, right?

**Comment by aficustree** [comment]
doesn't happen in the download request as xhr[i] is a GET, it seems if safari see the POST in shadow copies the entire variable scope and the deletes it when it's done. I think the theory is that window is a shared process. There's some spectre research notes on google too about 'site isolation' the window process and data leaking depending on the degree of per-process threading per site. I'm guessing safari has some exposure and that the theoretical attack vector is via a POST and not a GET.

**Comment by aficustree** [comment]
didn't think of that technique, will try it out when I get a moment.

**Comment by adolfintel** [comment]
Thanks. All you have to do is copy the postMessage that posts all the status as JSON at the end of the timer in ulTest.
If it works, I'll do the rest of the work myself.
Let me know asap ;)

**Comment by aficustree** [comment]
sorry, struggled to follow, I moved `ulStatus = ((speed * 8 * settings.overheadCompensationFactor) / (settings.useMebibits ? 1048576 : 1000000)).toFixed(2)` and put that into upload.onprogress() to no effect. Not sure if that was what you asked. As for that, no-change, but again, not sure if that'd work.

I did force IE11 mode and that *did* work

**Comment by adolfintel** [comment]
Ok, we have a solution, didn't expect that IE11 workaround to fix it, but ok.

What I meant was to copy lines 77-88 after line 422. Can you try that?

**Comment by aficustree** [comment]
I did, doesn't seem to work. So only the 'onprogress' event in CORS and when the method is POST has this behavior...so odd.

Could the postmessage within the onprogress callback be another CORS approved way of sharing the information? I tried just dumping the postmessage into the callback and that seems to result in a loop but I'm a little out of my expertise here.

**Comment by adolfintel** [comment]
I'm not sure. You could try putting the postMessage in the onProgress event and doing the calculations there instead of in the timer, but that will make the test use a ton of CPU on fast connections, which is something I want to avoid.

At this point, I feel more confident using the IE11 workaround. What's your user agent string?

**Comment by adolfintel** [comment]
Yes I could add a "small chunk size" parameter. Still, better than nothing.

Try adding this code after line 132:
```js
if (/^((?!chrome|android|crios|fxios).)*safari/i.test(ua)) {
        //Safari also needs the IE11 workaround but only for the MPOT version
        settings.forceIE11Workaround = true
      }
```

**Comment by aficustree** [comment]
seems to work. can you share more about that CPU impacts? The onprogress timer doesn't seem to fire all that frequently and the calculations just seem a couple FPU operations. I don't think it'd be over taxing for the processor.

**Comment by aficustree** [comment]
sure, I can do that, was going to try and add a check for a CORS header and only apply if the CORS header is detected. My copy of the repo has gotten quite misaligned to the main branch as I was adapting the mpot branch into a vuejs app and had made some fairly heavy modifications to the server selection logic and telemetry modules to comply with our use-case. As part of it I moved quite a bit around to make it all more modular and I see another PR that looks to have done the same. I'll submit all the changes after that PR is integrated.

**Comment by adolfintel** [comment]
@aficustree I'm still waiting for your PR. If you don't have time to do it it's not a problem, just let me know.
